### PR TITLE
Push filters down

### PR DESCRIPTION
**Pull request by nare469**
Closes #12 

This is also rebased off #11, so the older PR must be merged before this one.

### CODE REVIEW

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import ca.waterloo.dsg.graphflow.query.structuredquery.InClausePredicate;
+import ca.waterloo.dsg.graphflow.query.structuredquery.QueryPredicate;
+import ca.waterloo.dsg.graphflow.util.DataType;
+import ca.waterloo.dsg.graphflow.util.RuntimeTypeBasedComparator.ComparisonOperator;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+public class LiteralPredicateFactory {
+    public static Predicate<Integer> getLiteralFilterPredicate(List<QueryPredicate> queryPredicates) {
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java]: Add a blank line here.

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import ca.waterloo.dsg.graphflow.graph.VertexPropertyStore;
+import ca.waterloo.dsg.graphflow.query.structuredquery.ComparisonPredicate;
+import ca.waterloo.dsg.graphflow.query.structuredquery.InClausePredicate;
+import ca.waterloo.dsg.graphflow.query.structuredquery.QueryPredicate;
+import ca.waterloo.dsg.graphflow.util.DataType;
+import ca.waterloo.dsg.graphflow.util.RuntimeTypeBasedComparator.ComparisonOperator;
+
+import java.util.List;
+import java.util.function.Predicate;
+
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java]: class documentation missing.

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            } else {
+                result = result.and(predicate);
+            }
+        }
+        return result;
+    }
+
+    private static Predicate<Integer> getInClauseLiteralOnlyPredicate(InClausePredicate inClause) {
+        short propertyKey = TypeAndPropertyKeyStore.getInstance()
+            .mapStringPropertyKeyToShort(inClause.getLeftOperand().b);
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java]: '.mapStringPropertyKeyToShort(' should be on the previous line.

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        DataType dataType = TypeAndPropertyKeyStore.getInstance().getPropertyDataType(
+            inClause.getLeftOperand().b);
+
+        return (Integer prefix) -> {
+            String propertyValue = (String) VertexPropertyStore.getInstance().getProperty(prefix,
+                propertyKey);
+            for (String literal : inClause.getLiterals()) {
+                if (PredicateFactory.resolveTypesAndCheckEquality(dataType, propertyValue, literal,
+                    ComparisonOperator
+                    .EQUALS)) {
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java]: .EQUALS should be on the previous line.

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                }
+            }
+            return false;
+        };
+    }
+
+    private static Predicate<Integer> getComparativeClauseLiteralOnlyPredicate(
+        ComparisonPredicate predicate) {
+        short propertyKey = TypeAndPropertyKeyStore.getInstance()
+            .mapStringPropertyKeyToShort(predicate.getLeftOperand().b);
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/LiteralPredicateFactory.java]: '.mapStringPropertyKeyToShort(' should be on the previous line.

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/PredicateFactory.java [hunk 6]
```diff
@@ -49,7 +50,7 @@
         }
     }
 
-    private static boolean resolveTypesAndCheckEquality(DataType dataType, String value1,
+    public static boolean resolveTypesAndCheckEquality(DataType dataType, String value1,
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/filter/PredicateFactory.java]: should be package private.

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Extend.java [hunk 7]
```diff
@@ -1,26 +1,32 @@
...
 import java.util.Objects;
+import java.util.function.Predicate;
 
 /**
  * Given a set of output tuples of the previous operator, it extend the output to a new variable of
  * the query graph.
  */
 public class Extend extends StageOperator {
 
+    private Predicate<Integer> literalFilter;
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Extend.java]: literalFilter to -> literalVertexEntendedToPredicate

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Scan.java [hunk 8]
```diff
@@ -15,6 +17,8 @@
 public class Scan extends StageOperator {
 
     private short fromVertexTypeFilter;
+    private Predicate<Integer> literalFilterA;
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Scan.java]: rename to firstVertexLiteralPredicate

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Scan.java [hunk 9]
```diff
@@ -15,6 +17,8 @@
...
+    private Predicate<Integer> literalFilterB;
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Scan.java]: rename to secondVertexLiteralPredicate

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Scan.java [hunk 10]
```diff
@@ -26,9 +30,12 @@
      * is ignored.
      */
     public Scan(List<EdgeIntersectionRule> intersectionRules, short fromVertexTypeFilter,
-        short toVertexTypeFilter) {
+        short toVertexTypeFilter, Predicate<Integer> literalFilterA,
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Scan.java]: rename params based on fields names.

File: src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Extend.java [hunk 11]
```diff
@@ -64,14 +70,16 @@ public void append(int[][] prefixes) {
...
             for (int j = 0; j < extensions.getSize(); j++) {
-                int[] newPrefix = new int[prefix.length + 1];
-                System.arraycopy(prefix, 0, newPrefix, 0, prefix.length);
-                newPrefix[newPrefix.length - 1] = extensions.get(j);
-                newPrefixes[newPrefixCount++] = newPrefix;
-                // Append the prefixes to the next operator in size StageOperator.BATCH_SIZE.
-                if (BATCH_SIZE == newPrefixCount) {
-                    append(newPrefixes, BATCH_SIZE);
-                    newPrefixCount = 0;
+                if (literalFilter.test(extensions.get(j))) {
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/operator/join/Extend.java]: Do the proper changes such that this is set to null if no predicates exist which is faster then function calls that return true.

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java [hunk 12]
```diff
@@ -129,6 +130,30 @@ void orderRemainingVariables(List<String> orderedVariables) {
...
+            queryPredicatesByVariable.get(predicate.getLeftOperand().a).add(predicate);
+        }
+        List<Predicate<Integer>> orderedPredicates = new ArrayList<>();
+        for (String variable : orderedVariables) {
+            List<QueryPredicate> predicatesForVariable = queryPredicatesByVariable.get(variable);
+            if (null != predicatesForVariable) {
+                orderedPredicates.add(LiteralPredicateFactory.getLiteralFilterPredicate
+                    (predicatesForVariable));
+            } else {
+                // TODO(Naren): Fix this. This is just a temporary fix. Make it cleaner
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java]: yee! simply give it a null predicate.

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java [hunk 13]
```diff
@@ -168,6 +193,8 @@ public QueryPlan plan() {
         String fromVertexTypeFilter = null;
         String toVertexTypeFilter = null;
         // Start from the second variable to create the first stage.
+        List<Predicate<Integer>> orderedPredicates = getOrderedPredicatesFromVariables
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java]: move this to line 192

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java [hunk 14]
```diff
@@ -192,12 +219,13 @@ public QueryPlan plan() {
             if (1 == i) {
                 currentStageOperator = new Scan(stage, TypeAndPropertyKeyStore.getInstance().
                     mapStringTypeToShort(fromVertexTypeFilter), TypeAndPropertyKeyStore.
-                    getInstance().mapStringTypeToShort(toVertexTypeFilter));
+                    getInstance().mapStringTypeToShort(toVertexTypeFilter), orderedPredicates.get
+                    (0), orderedPredicates.get(1));
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java]: move the first char '(' to previous line if possible else 'get' should go back in line.

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java [hunk 15]
```diff
@@ -168,6 +193,8 @@ public QueryPlan plan() {
         String fromVertexTypeFilter = null;
         String toVertexTypeFilter = null;
         // Start from the second variable to create the first stage.
+        List<Predicate<Integer>> orderedPredicates = getOrderedPredicatesFromVariables
+            (orderedVariables);
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java]: move first char '(' to previous line if not possible then the whole function name should go to next line.

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java [hunk 16]
```diff
@@ -285,7 +313,7 @@ AbstractOperator getNextOperator(List<String> orderedVertexVariablesBeforeProjec
         }
 
         // Construct the {@code Filter} operator if needed.
-        if (!structuredQuery.getQueryPredicates().isEmpty()) {
+        if (!structuredQuery.getNonVertexLiteralPredicates().isEmpty()) {
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/OneTimeMatchQueryPlanner.java]: rename this to getEdgeLiteralAndNonLiteralPredicates()

File: src/main/java/ca/waterloo/dsg/graphflow/query/structuredquery/StructuredQuery.java [hunk 17]
```diff
@@ -78,7 +77,48 @@ public void addVariable(QueryVariable queryVariable) {
      * @param queryPredicate The {@link QueryPredicate} to be added.
      */
     public void addQueryPredicate(QueryPredicate queryPredicate) {
-        this.queryPredicates.add(queryPredicate);
+        switch (queryPredicate.getPredicateType()) {
+            case COMPARATIVE_CLAUSE_PROPERTY_KEY_AND_LITERAL_OPERANDS:
+            case IN_CLAUSE_ONLY_LITERALS:
+                String leftOperandName = queryPredicate.getLeftOperand().a;
+                for (QueryVariable variable : queryVariables) {
+                    if (leftOperandName == variable.getVariableName()) {
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/structuredquery/StructuredQuery.java]: == -> .equals()

File: src/main/java/ca/waterloo/dsg/graphflow/query/structuredquery/StructuredQuery.java [hunk 18]
```diff
@@ -78,7 +77,48 @@ public void addVariable(QueryVariable queryVariable) {
...
+                        vertexLiteralPredicates.add(queryPredicate);
+                        return;
+                    }
+                }
+                nonVertexLiteralPredicates.add(queryPredicate);
+                return;
+            case COMPARATIVE_CLAUSE_TWO_PROPERTY_KEY_OPERANDS:
+            case IN_CLAUSE_VARIABLES_AND_LITERALS:
+                nonVertexLiteralPredicates.add(queryPredicate);
+                return;
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/structuredquery/StructuredQuery.java]: why return and not break?

File: src/main/java/ca/waterloo/dsg/graphflow/query/structuredquery/StructuredQuery.java [hunk 19]
```diff
@@ -78,7 +77,48 @@ public void addVariable(QueryVariable queryVariable) {
...
+            case IN_CLAUSE_ONLY_LITERALS:
+                String leftOperandName = queryPredicate.getLeftOperand().a;
+                for (QueryVariable variable : queryVariables) {
+                    if (leftOperandName == variable.getVariableName()) {
+                        vertexLiteralPredicates.add(queryPredicate);
+                        return;
+                    }
+                }
+                nonVertexLiteralPredicates.add(queryPredicate);
+                return;
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/structuredquery/StructuredQuery.java]: why return and not break?

Comment (nare469) [reply src/main/java/ca/waterloo/dsg/graphflow/query/structuredquery/StructuredQuery.java]: Trying to keep it uniform with the previous thing. That needs to return.

File: src/main/java/ca/waterloo/dsg/graphflow/query/structuredquery/StructuredQuery.java (hunk 18 above)
Comment (nare469) [reply src/main/java/ca/waterloo/dsg/graphflow/query/structuredquery/StructuredQuery.java]: see previous comment

**Comment by nare469** [comment]
Closing this now that #11 has been merged

**Pull request closed without merging**